llvm/include/llvm/CodeGen/BasicTTIImpl.h

```diff
@@ -1910,9 +1906,15 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
           BinaryOperator::LShr, RetTy, CostKind, OpInfoY,
           {OpInfoZ.Kind, TTI::OP_None});
       // Non-constant shift amounts requires a modulo.
-      if (!OpInfoZ.isConstant())
-        Cost += thisT()->getArithmeticInstrCost(BinaryOperator::URem, RetTy,
-                                                CostKind, OpInfoZ, OpInfoBW);
+      if (!OpInfoZ.isConstant()) {
+        Cost += isPowerOf2_32(RetTy->getScalarSizeInBits())
+                    ? thisT()->getArithmeticInstrCost(
+                          BinaryOperator::And, RetTy, CostKind, OpInfoZ,
+                          {TTI::OK_UniformConstantValue, TTI::OP_None})
+                    : thisT()->getArithmeticInstrCost(
+                          BinaryOperator::URem, RetTy, CostKind, OpInfoZ,
+                          {TTI::OK_UniformConstantValue, TTI::OP_None});
```

Comment:
```suggestion
        Cost += thisT()->getArithmeticInstrCost(
                          isPowerOf2_32(RetTy->getScalarSizeInBits()) ? BinaryOperator::And : BinaryOperator::URem, RetTy, CostKind, OpInfoZ,
                          {TTI::OK_UniformConstantValue, TTI::OP_None});
```

Can make the opcode the conditional part